src/awx_plugins/credentials/aws_assumerole.py

```diff
@@ -0,0 +1,179 @@
+    :type role_arn: str
+    :param external_id: The external ID received from AWS.
+    :type external_id: int
+    :returns: The credentials received from AWS.
+    :rtype: dict
+    :raises ValueError: If the client response is bad.
+    """
+    connection: STSClient = boto3.client(
+        service_name='sts',
+        # The following EE creds are read from the env if they are not passed:
+        aws_access_key_id=access_key,  # defaults to `None` in the lib
+        aws_secret_access_key=secret_key,  # defaults to `None` in the lib
+    )
+    try:
+        response: AssumeRoleResponseTypeDef = connection.assume_role(
+            RoleArn=role_arn,
+            RoleSessionName='AAP_AWS_Role_Session1',
+            ExternalId=external_id,
+        )
+    except ClientError as client_err:
+        raise ValueError(
+            f'Got a bad client response from AWS: {client_err.message}.',
+        ) from client_err
+
+    return response.get('Credentials', {})
+
+
+def aws_assumerole_backend(
+        access_key: str | None,
+        secret_key: str | None,
+        role_arn: str | None,
+        external_id: int,
+        identifier: str,
+) -> dict:
+    """Contact AWS to assume a given role for the user.
+
+    :param access_key: The AWS access key ID.
+    :type access_key: str
+    :param secret_key: The AWS secret access key.
+    :type secret_key: str
+    :param role_arn: The ARN received from AWS.
+    :type role_arn: str
+    :param external_id: The external ID received from AWS.
+    :type external_id: int
+    :param identifier: The identifier to fetch from the assumed role.
+    :type identifier: str
+    :raises ValueError: If the identifier is not found.
+    :returns: The identifier fetched from the assumed role.
+    :rtype: dict
+    """
+    # Generate a unique SHA256 hash for combo of user access key and ARN
+    # This should allow two users requesting the same ARN role to have
+    # separate credentials, and should allow the same user to request
+    # multiple roles.
+    credential_key_hash = hashlib.sha256(
+        (str(access_key or '') + role_arn).encode('utf-8'),
+    )
+    credential_key = credential_key_hash.hexdigest()
+
+    credentials = _aws_cred_cache.get(credential_key, None)
+
+    # If there are no credentials for this user/ARN *or* the credentials
+    # we have in the cache have expired, then we need to contact AWS again.
+    creds_expired = (
+        (creds_expire_at := credentials.get('Expiration')) and
+        creds_expire_at < datetime.now(credentials['Expiration'].tzinfo)
+    )
+    if creds_expired:
+
+        credentials = aws_assumerole_getcreds(
+            access_key, secret_key, role_arn, external_id,
+        )
+
+        _aws_cred_cache[credential_key] = credentials
+
+    credentials = _aws_cred_cache.get(credential_key, {})
+
+    try:
+        return credentials[identifier]
+    except KeyError as key_err:
+        raise ValueError(
```

Comment:
The same concern goes for this exception. Although I imagine that awx expects a `ValueError` so we can't change it w/o defining the interface first...

Reply:
that is my thought as well, looking at the actual error message makes it seem like this should be dev focused since we are the ones going and getting this actual value with the user inputs but could be helpful for debugging if someone hits this. kinda leaves a good flag as where things could have fallen apart?

Reply:
Technically, all the exceptions are targeting the devs, so are the tracebacks. But if awx uses this somehow, we can't change it to a `LookupError`, I think.